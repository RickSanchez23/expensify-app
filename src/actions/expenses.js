import database from '../firebase/firebase';

export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense,
});

export const addExpenseAsync = (expenseData) => {
    const {amount = 0, description = "", note = "", createAt = 0} = expenseData;
    return (dispatch) => {
        const expense = {amount, description, note, createAt}
        return database.ref('expenses')
            .push(expense)
            .then(ref => {
                dispatch(addExpense({
                    ...expense,
                    id: ref.key
                }));
            }).catch(err => {
            });
    }
}

export const deleteExpense = (itemId) => ({
    type: "REMOVE_EXPENSE",
    itemId,
});

export const editExpense = (itemId, updates) => ({
    type: "EDIT_EXPENSE",
    itemId,
    updates,
});

export const setExpenses = (expenses) => {
    return {
        type: 'SET_EXPENSES',
        expenses

    }
}

export const setExpensesAsync = () => {

    return (dispatch) => {
        return database.ref('expenses').once('value').then(snapshot => {
            const expensesList = [];
            snapshot.forEach(snapshotChild => {
                expensesList.push({
                    id : snapshotChild.key ,
                    ...snapshotChild.val()
                });
            });
            dispatch(setExpenses(expensesList));
        });

    }
}
