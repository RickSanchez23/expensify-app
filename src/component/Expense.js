import React from "react";
import { NavLink } from "react-router-dom";

const Expense = ({ amount, description, createAt, id}) => {
  return (
    <div>
      <NavLink to={"/edit/" + id}>
        <h3>{description}</h3>
      </NavLink>
      <p>{amount}$ - {createAt}</p>
    </div>
  );
};

export default Expense;
