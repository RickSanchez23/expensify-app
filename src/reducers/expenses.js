const expenseReducerDefaultState = [];
export default (state = expenseReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];

    case "REMOVE_EXPENSE":
      return state.filter((item) => item.id !== action.itemId);

    case "EDIT_EXPENSE":
      return state.map((item) => {
        if (item.id === action.itemId) {
          return { ...item, ...action.updates };
        } else {
          return item;
        }
      });

    case "SET_EXPENSES" :
      return action.expenses

    default:
      return state;
  }
};
