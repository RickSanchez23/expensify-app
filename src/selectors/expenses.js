export const getVisibleExpenses = (
  expenses,
  { text, startDate, endDate, sortBy }
) => {
  return expenses
    .filter((item) => {
      const textFilterMatch = item.description
        .toLowerCase()
        .includes(text.toString());
      const startDateMatch =
        typeof startDate !== "number" || item.createAt >= startDate;
      const endDateMatch =
        typeof endDate !== "number" || item.createAt <= endDate;
      return textFilterMatch && startDateMatch && endDateMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createAt < b.createAt ? 1 : -1;
      } else {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};
