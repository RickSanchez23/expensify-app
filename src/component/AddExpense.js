import React from "react";
import ExpenseFrom from "./ExpesnseForm";
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";

export class AddExpense extends React.Component {
    onExpenseFromSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push("/");
    };

    render() {
        return (
            <div>
                <h5>Add New Expense</h5>
                <ExpenseFrom onSubmit={this.onExpenseFromSubmit} />
            </div>
        );
    }
}
const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(AddExpense);
