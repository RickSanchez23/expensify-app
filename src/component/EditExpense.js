import React from "react";
import ExpenseFrom from "./ExpesnseForm";
import { connect } from "react-redux";
import { deleteExpense, editExpense } from "../actions/expenses";

export class EditExpense extends React.Component {
  onDeleteClick = () => {
    this.props.deleteExpense(this.props.expense.id);
    this.props.history.push("/");
  }

  onSubmitExpenseFrom = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  }

  render() {
    return (
      <div>
        <h5>Edit Expense</h5>
        <ExpenseFrom
          expense={this.props.expense}
          onSubmit={this.onSubmitExpenseFrom}
        />
        <button onClick={this.onDeleteClick}>delete</button>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  deleteExpense: (itemId) => dispatch(deleteExpense(itemId)),
  editExpense: (itemId, expense) => dispatch(editExpense(itemId, expense)),
});

const mapStateToProps = (state, props) => {
  return {
    expense: state.expense.find((item) => item.id === props.match.params.id),
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);
