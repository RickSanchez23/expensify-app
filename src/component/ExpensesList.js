import React from "react";
import { connect } from "react-redux";
import Expense from "./Expense";
import { getVisibleExpenses } from "../selectors/expenses";

export const ExpensesList = (props) => (
  <div>
    {
      !props.expenses || props.expenses.length === 0 ? (
          <p>there is no expenses! </p>
      ) : (
          props.expenses.map((expense, index) => (
            <Expense {...expense} key={index} />
            ))
      )
    }

  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: getVisibleExpenses(state.expense, state.filters),
  };
};

export default connect(mapStateToProps)(ExpensesList);
