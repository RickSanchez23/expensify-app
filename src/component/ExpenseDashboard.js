import React from "react";
import ExpensesList from "./ExpensesList";
import ExpenseTextFilterFiled from "./ExpenseTextFilterFiled";

const ExpenseDashboard = () => (
  <div>
    <ExpenseTextFilterFiled />
    <ExpensesList />
  </div>
);

export default ExpenseDashboard;
