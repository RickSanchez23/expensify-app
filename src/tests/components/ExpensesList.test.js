import React from "react";
import { shallow } from "enzyme";
import { ExpensesList } from "../../component/ExpensesList";
import expenses from "../fixtures/expenses";

test("should render ExpenseList", () => {
  const wrapper = shallow(<ExpensesList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseList with null data", () => {
  const wrapper = shallow(<ExpensesList />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseList with empty data", () => {
  const wrapper = shallow(<ExpensesList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
