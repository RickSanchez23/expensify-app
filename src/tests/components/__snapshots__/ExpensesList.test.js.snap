// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList 1`] = `
<div>
  <Expense
    amount={310}
    createAt={-345600000}
    description="Rent"
    id="1"
    key="0"
  />
  <Expense
    amount={2}
    createAt={-172800000}
    description="Coffee"
    id="2"
    key="1"
  />
  <Expense
    amount={12}
    createAt={432000000}
    description="Gas Bill"
    id="3"
    key="2"
  />
  <Expense
    amount={8}
    createAt={0}
    description="Food"
    id="4"
    key="3"
  />
</div>
`;

exports[`should render ExpenseList with empty data 1`] = `
<div>
  <p>
    there is no expenses! 
  </p>
</div>
`;

exports[`should render ExpenseList with null data 1`] = `
<div>
  <p>
    there is no expenses! 
  </p>
</div>
`;
